Source/ProgressionSystemRuntime/Private/Data/PSWorldSubsystem.cpp

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Valerii Rotermel & Yevhenii Selivanov
+
+#include "Data/PSWorldSubsystem.h"
+
+#include "ProgressionSystemRuntimeModule.h"
+#include "Controllers/MyPlayerController.h"
+#include "Data/PSDataAsset.h"
+#include "Data/PSSaveGameData.h"
+#include "Kismet/GameplayStatics.h"
+#include "LevelActors/PlayerCharacter.h"
+#include "MyDataTable/MyDataTable.h"
+#include "MyUtilsLibraries/UtilsLibrary.h"
+#include "UtilityLibraries/MyBlueprintFunctionLibrary.h"
+
+#include UE_INLINE_GENERATED_CPP_BY_NAME(PSWorldSubsystem)
+
+// Returns this Subsystem, is checked and wil crash if can't be obtained
+UPSWorldSubsystem& UPSWorldSubsystem::Get()
+{
+	const UWorld* World = UUtilsLibrary::GetPlayWorld();
+	checkf(World, TEXT("%s: 'World' is null"), *FString(__FUNCTION__));
+	UPSWorldSubsystem* ThisSubsystem = World->GetSubsystem<ThisClass>();
+	checkf(ThisSubsystem, TEXT("%s: 'ProgressiorSubsystem' is null"), *FString(__FUNCTION__));
+	return *ThisSubsystem;
+}
+
+// Returns this Subsystem, is checked and wil crash if can't be obtained
+UPSWorldSubsystem& UPSWorldSubsystem::Get(const UObject& WorldContextObject)
+{
+	const UWorld* World = GEngine->GetWorldFromContextObjectChecked(&WorldContextObject);
+	checkf(World, TEXT("%s: 'World' is null"), *FString(__FUNCTION__));
+	UPSWorldSubsystem* ThisSubsystem = World->GetSubsystem<ThisClass>();
+	checkf(ThisSubsystem, TEXT("%s: 'ProgressiorSubsystem' is null"), *FString(__FUNCTION__));
+	return *ThisSubsystem;
+}
+
+// Returns the data asset that contains all the assets of Progression System game feature
+const UPSDataAsset* UPSWorldSubsystem::GetPSDataAsset() const
+{
+	return PSDataAssetInternal.LoadSynchronous();
+}
+
+void UPSWorldSubsystem::SetProgressionSystemComponent(UPSHUDComponent* MyProgressionSystemComponent)
+{
+	checkf(MyProgressionSystemComponent, TEXT("%s: My progression system component is null"), *FString(__FUNCTION__));
+	PSHUDComponentInternal = MyProgressionSystemComponent;
+}
+
+void UPSWorldSubsystem::OnWorldBeginPlay(UWorld& InWorld)
+{
+	Super::OnWorldBeginPlay(InWorld);
+
+	// Listen events on player type changed and Character spawned
+	if (APlayerCharacter* MyPlayerCharacter = UMyBlueprintFunctionLibrary::GetLocalPlayerCharacter())
+	{
+		MyPlayerCharacter->OnPlayerTypeChanged.AddUniqueDynamic(this, &ThisClass::OnPlayerTypeChanged);
+	}
+	else if (AMyPlayerController* MyPC = UMyBlueprintFunctionLibrary::GetLocalPlayerController())
+	{
+		MyPC->GetOnNewPawnNotifier().AddUObject(this, &ThisClass::OnCharacterPossessed);
+	}
+
+	// requires to be executed to guarntee that data asset will be loaded from the ini file
+	// otherwise the data asset file will not be loaded at fresh (heaviest) load of UE
+	PSDataAssetInternal = GetPSDataAsset();
+	if (PSDataAssetInternal)
+	{
+		ProgressionDataTableInternal = PSDataAssetInternal->GetProgressionDataTable();
+		LoadGameFromSave();
+	}
+	else
+	{
+		UE_LOG(LogProgressionSystem, Warning, TEXT("PSDataAssetInternal null"))
```

Comment:
1. Please always prefer `ensureMsgf` over theUE_LOG:
```
if (ensureMsgf(PSDataAssetInternal, TEXT("ASSERT: some message")
{
    ProgressionDataTableInternal = ...
}
```
It does the same as log, but also makes assertion in debugger. While logs are very often ignored by developers, it's hard to ignore `ensure`.

Alternatively, you can use `checkf` instead of `ensureMsgf` if you want to crash the engine: only for fatal errors in logic.

I use `ensureMsgf` where there is chance something setup wrong in data assets and `checkf` when my logic goes very wrong (like no created subsystem when it always has to be).

2. Once you remove this, please also get rid of `include ProgressionSystemRuntimeModule.h`

Reply:
done, also additional round of cleaning headers performed